ast/interning.go

```diff
@@ -0,0 +1,561 @@
+// Copyright 2024 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package ast
+
+import "strconv"
+
+// NOTE! Great care must be taken **not** to modify the terms returned
+// from these functions, as they are shared across all callers.
+
+var (
+	booleanTrueTerm  = &Term{Value: Boolean(true)}
+	booleanFalseTerm = &Term{Value: Boolean(false)}
+
+	// since this is by far the most common negative number
+	minusOneTerm = &Term{Value: Number("-1")}
+)
+
+func InternedBooleanTerm(b bool) *Term {
+	if b {
+		return booleanTrueTerm
+	}
+
+	return booleanFalseTerm
+}
+
+// InternedIntNumberTerm returns a term with the given integer value. The term is
+// cached between -1 to 512, and for values outside of that range, this function
```

Comment:
Not really, mostly just from most common numbers I saw Regal process (which is never above 512). I don't have any preference here, and I don't mind increasing it if you think that's a good idea. 

Dynamic caching would be nice, but it would also mean we'd need to consider if invalidation is needed, and so on. Not against that either, but not something I felt like I had the time to spend on currently.

Reply:
We can add a more dynamic approach later if needed. This feels fine to me. Not a strong opinion on the range we cache.